packages/core/snapshot/snapshotwrite.go

```diff
@@ -0,0 +1,184 @@
+			if err != nil {
+				return err
+			}
+			chunksOutputWithMetadata = make([]*ledger.OutputWithMetadata, 0)
+		}
+	}
+
+	// write epochDiffs
+	epochDiffs, err := epochDiffsProd()
+	if err != nil {
+		return err
+	}
+
+	typeSet := new(serix.TypeSettings)
+	bytes, err := serix.DefaultAPI.Encode(context.Background(), epochDiffs, serix.WithTypeSettings(typeSet.WithLengthPrefixType(serix.LengthPrefixTypeAsUint32)), serix.WithValidation())
+	if err != nil {
+		return err
+	}
+	if err := writeFunc(fmt.Sprintf("diffEpoch"), append(bytes, delimiter...)); err != nil {
+		return err
+	}
+
+	// seek back to the file position of the outputWithMetadata counter
+	if _, err := writeSeeker.Seek(0, io.SeekStart); err != nil {
+		return fmt.Errorf("unable to seek to LS counter placeholders: %w", err)
+	}
+	if err := writeFunc(fmt.Sprintf("outputWithMetadata counter %d", outputWithMetadataCounter), outputWithMetadataCounter); err != nil {
+		return err
+	}
+	header.OutputWithMetadataCount = outputWithMetadataCounter
+
+	return nil
+}
+
+// NewLedgerOutputWithMetadataProducer returns a OutputWithMetadataProducerFunc that provide OutputWithMetadatas from the ledger.
+func NewLedgerOutputWithMetadataProducer(l *ledger.Ledger) OutputWithMetadataProducerFunc {
+	prodChan := make(chan interface{})
+
+	go func() {
+		l.ForEachAcceptedUnSpentOutputWithMetadata(func(o *ledger.OutputWithMetadata) {
+			prodChan <- o
+		})
+
+		close(prodChan)
+	}()
+
+	binder := producerFromChannels(prodChan)
+	return func() *ledger.OutputWithMetadata {
+		obj := binder()
+		if obj == nil {
+			return nil
+		}
+		return obj.(*ledger.OutputWithMetadata)
+	}
+}
+
+func writeSnapshotHeader(writeSeeker io.WriteSeeker, header *ledger.SnapshotHeader) error {
+	writeFunc := func(name string, value any, offsetsToIncrease ...*int64) error {
+		return writeFunc(writeSeeker, name, value, offsetsToIncrease...)
+	}
+
+	if err := writeFunc(fmt.Sprintf("outputWithMetadata counter %d", header.OutputWithMetadataCount), header.OutputWithMetadataCount); err != nil {
+		return err
+	}
+
+	if err := writeFunc(fmt.Sprintf("fullEpochIndex %d", header.FullEpochIndex), header.FullEpochIndex); err != nil {
+		return err
+	}
+
+	if err := writeFunc(fmt.Sprintf("diffEpochIndex %d", header.DiffEpochIndex), header.DiffEpochIndex); err != nil {
+		return err
+	}
+
+	data, err := header.LatestECRecord.Bytes()
+	if err != nil {
+		return err
+	}
+
+	if err := writeFunc("latestECRecord", append(data, delimiter...)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func producerFromChannels(prodChan <-chan interface{}) func() interface{} {
+	return func() interface{} {
+		select {
+		case obj, ok := <-prodChan:
+			if !ok {
+				return nil
+			}
+			return obj
+		}
+	}
+}
+
+func increaseOffsets(amount int64, offsets ...*int64) {
+	for _, offset := range offsets {
+		*offset += amount
+	}
+}
+
+func writeFunc(writeSeeker io.WriteSeeker, variableName string, value any, offsetsToIncrease ...*int64) error {
```

Comment:
Do we ever call this function with some `offsetsToIncrease` ?

Reply:
No, but I keep it there in case we need that in the future.

Reply:
Let's remove it then, otherwise it seems that we are actually using it somewhere to increase a counter.